core/modules/filter/filter.pages.inc

```diff
@@ -491,28 +491,146 @@ function filter_format_editor_dialog_save($form, &$form_state) {
-      $value = '/' . $value;
-    }
+  // Allow empty values, since this will not result in a link tag anyway.
+  if ($value === '') {
+    form_set_value($element, $value, $form_state);
+    return;
   }
 
+  // Cleanup the URL so it's more likely to pass validation.
+  $value = _filter_cleanup_url($value);
+
   form_set_value($element, $value, $form_state);
 
+  // Flag if it has any dangerous protocols.
+  $cleaned_url = backdrop_strip_dangerous_protocols($value);
+  $dangerous_protocols = $cleaned_url !== $value ? TRUE : FALSE;
+
   // Unlike #type = 'url' validation, we allow both relative and absolute paths.
-  if ($value !== '' && !valid_url($value, TRUE) && !valid_url($value, FALSE)) {
+  if ($dangerous_protocols || (!valid_url($value, TRUE) && !valid_url($value, FALSE))) {
     form_error($element, t('The URL %url is not valid.', array('%url' => $value)));
   }
 }
 
+/**
+ * Clean up URL for validation.
+ *
+ * Attempts to create a URL that will pass validation: adds a protocol if it's
+ * missing; urlencodes an anchor link; attempts to find the proper multilingual
+ * URL; and urlencodes any internal Backdrop path.
+ *
+ * @param string $url
+ *
+ * @return string
+ *   A cleaned-up URL that is either a properly formatted external URL or an
+ *   internal path.
+ *
+ * @see _filter_format_editor_link_url_validate()
+ */
+function _filter_cleanup_url($url) {
+  // Will add a default protocol if missing on an external url.
+  $url = _filter_create_valid_external_url($url);
+
+  // If it's an external URL we can skip the rest of the clean up.
+  if (_filter_url_is_external($url)) {
+    return $url;
+  }
+
+  // Encode an anchor as it might have been typed in with unicode.
+  if (substr($url, 0, 1) == '#' && preg_match('/^(#)([\S]+)$/', $url, $matches)) {
+    return $matches['1'] . backdrop_encode_path($matches['2']);
+  }
+
+  // Break up the components so we can build up an aliased internal path.
+  $url_parts = backdrop_parse_url($url);
+
+  // Lookup valid internal paths if multilingual
+  if (module_exists('locale') && language_multilingual()) {
+    if ($language = _filter_lookup_locale_path($url)) {
+      $url_parts['language'] = $language;
+    }
+  }
+
+  return url($url_parts['path'], $url_parts);
+}
+
+/**
+ * Forms a valid external URL if possible.
+ *
+ * Trims whitespace and automatically adds an http:// to addresses that
+ * appear to be external URLs but have no specified protocol.
+ *
+ * @param string $url
+ *   The URL entered by the user.
+ *
+ * @return string
+ *   The URL with the protocol prepended if needed.
+ *
+ * @see link_cleanup_url().
+ */
+function _filter_create_valid_external_url($url) {
+  // Skip if it's already a valid external URL.
+  if (_filter_url_is_external($url)) {
+    return $url;
+  }
+
+  // Should there be a protocol? Add an automatic http:// if it starts with a
+  // domain name.
+  $domain_match = preg_match('/^(([a-z0-9]([a-z0-9\-_]*\.){1,63})([a-z]{2,63}))/i', $url);
```

Comment:
Yet another regex... I avoided regex in my PR by intention, and used parse_url() instead. But you seem to be huge fan of regular expressions...

Reply:
You've assumed I'm a "huge fan" of regular expressions, but I think you are misled. I'm a fan of using what's available and I found parts of functions in Link module and in core useful. Both are using regular expressions in a number of places. If you search core for the regex strings above you'll see. Plus I've put it in the docblock so you can see where code comes from.

I don't find regex as intimidating anymore since I use sites like https://regex101.com/ a lot for seeing how the regex will act on example strings.

Reply:
To clarify: I don't find regex intimidating, but still avoid it where not necessary.
Above regex is (IMO) easy to avoid - why not using parse_url() to get the parts? That's not any worse and so much easier to read.

Reply:
`parse_url` doesn't work here with `//domain.tld`. It will only tell us if there's a host with no scheme. It totally ignores the //.

Reply:
Can you explain a bit more, please?

I don't fully get what's happening here. You claim to automatically add a scheme, but that doesn't actually happen.
I get the same result for '//domain.tld" and 'http://domain.tld', so I suspect, this doesn't do what you expect. And I find the comments pretty misleading.

Simplified code example:
```
<?php
$url1 = '//domain.tld';
$url2 = 'https://domain.tld';

$domain_match1 = preg_match('/^(([a-z0-9]([a-z0-9\-_]*\.){1,63})([a-z]{2,63}))/i', $url1);
$domain_match2 = preg_match('/^(([a-z0-9]([a-z0-9\-_]*\.){1,63})([a-z]{2,63}))/i', $url2);

var_dump($domain_match1);
var_dump($domain_match2);
?> 
```

Reply:
This works like it does in `link_cleanup_url()`. Neither `//domain.tld` nor `https://domain.tld` should be matched by this so it's fine that it has the same result. It *will* however match on `domain.tld` or `domain.ca.tld` which is what we want. If it has a match then we know we can safely add an "http://" to create an external URL.

`//domain.tld` passes `valid_url()` so it's not trying to rewrite it.

Reply:
@herbdool many thanks for your explanation. It wasn't obvious for me, what this is supposed to do. Yes, in this case neither parse_url nor filter_var are helpful. You can close this as "resolved".